### PR TITLE
feat: support migrations during version upgrades

### PR DESCRIPTION
**Pull request by cmwaters**
Closes: #3105

### CODE REVIEW

File: app/app.go
Comment (rootulp) [inline app/app.go:393]: [question] why is v2.Version used as both the second and third param? Is it because the upgrade module only supports v2 and isn't enabled in v1?

See other comment but Go docs on what the `fromVersion` and `toVersion` describe would help.

File: app/app.go
Comment (rootulp) [inline app/app.go:544]: [question] why is `version` used for both `fromVersion` and `toVersion`? If we're upgrading, from a version to a different version, I would expect these two versions to be distinct.

Can a clarifying comment be added?

File: app/module/configurator.go
Comment (rootulp) [inline app/module/configurator.go:48]: does this mean no migrations can occur from v0 -> v1? Does other code already assume version starts at v1?

File: app/module/configurator.go
Comment (rootulp) [inline app/module/configurator.go:68]: given the logic `fromVersion == toVersion`, I expect the other line of code to never run any migrations

```go
	return app.mm.RunMigrations(ctx, app.configurator, version, version)
```

File: app/module/module.go
Comment (rootulp) [inline app/module/module.go:54]: [uber nit] can the conditional logic and the error message match so that it's easier to mentally verify 

```suggestion
			return nil, sdkerrors.ErrLogic.Wrapf("fromVersion can not be greater than toVersion for module %s", module.module.Name())
```

File: app/module/module.go
Comment (rootulp) [inline app/module/module.go:96]:
```suggestion
// SetOrderBeginBlockers sets the order of begin-blocker calls
```

File: app/module/module.go
Comment (rootulp) [inline app/module/module.go:102]:
```suggestion
// SetOrderEndBlockers sets the order of end-blocker calls
```

File: app/module/module.go
Comment (rootulp) [inline app/module/module.go:231]: [optional] to disambiguate from the error below

```suggestion
		return sdkerrors.ErrInvalidVersion.Wrapf("fromVersion %d not supported", fromVersion)
```

File: app/module/module.go
Comment (rootulp) [inline app/module/module.go:235]: [optional] to disambiguate from the error above

```suggestion
		return sdkerrors.ErrInvalidVersion.Wrapf("toVersion %d not supported", toVersion)
```

File: app/module/module.go
Comment (rootulp) [inline app/module/module.go:80]: [tangent] I don't think we need to do this at the same time as this PR but I wonder if all these could use the same order: https://github.com/celestiaorg/celestia-app/issues/3052.

**Review by rootulp — COMMENTED** [review]
Kinda makes sense to me. My takeaway is this introduces a new versioned module type that is aware of the version range it supports. That seems reasonable. The logic around running migrations also seems sound so no big blocking feedback

**Comment by cmwaters** [comment]
>  My takeaway is this introduces a new versioned module type that is aware of the version range it supports

Yeah it's basically that and it multiplexes based on the version it expects to be on

File: app/module/module.go
Comment (cmwaters) [reply app/module/module.go:80]: Yeah I was merely copying across from the SDK

File: app/module/configurator.go
Comment (cmwaters) [reply app/module/configurator.go:48]: Correct. Yes we started on v1 and there are assumptions that v0 doesn't exist and implies that the version has forgotten to be set

File: app/module/configurator.go
Comment (cmwaters) [reply app/module/configurator.go:68]: Yeah sorry, I just patched this up quickly without correctly implementing it

File: app/module/configurator.go
Comment (rootulp) [reply app/module/configurator.go:48]: thanks! 

[tangent] Is there already an assertion somewhere that fails on chain start up if a chain is initialized with v0? Asking because I haven't seen an assertion like that but I expect it exists _somewhere_.

Comment (cmwaters) [reply app/module/configurator.go:48]: Yes I believe I've added it. I'll double check

File: app/app.go
Comment (cmwaters) [reply app/app.go:544]: Changed it

File: app/module/module.go
Comment (evan-forbes) [inline app/module/module.go:245]: > // TODO: handle the case where a module is no longer supported (i.e. removed from the state machine)

Moving this todo to an issue

**Review by evan-forbes — DISMISSED** [review]
this LGTM

the mocks are great essential unit tests, and this might be planned and I just missed it, but it seems v2 blocking to have some test that will break if things outside of the Manager change. Similar to #3137, if there are other things that are necessary for the complete integration, can we make a point to document them before merging and then discuss them in the next v2 sync? cc @rootulp

File: app/module/module.go
Comment (cmwaters) [reply app/module/module.go:39]: It's meant as a range. The same module could be used over several app versions

**Review by evan-forbes — APPROVED** [review]

File: app/app.go
Comment (rootulp) [inline app/app.go:555]: I don't understand this comment, can it be rephrased? The consensus params and the validator set seem orthogonal. 

Also the validator set doesn't seem applicable to the following three lines so if it's not relevant in this context, proposal to delete that portion of the comment.

**Review by rootulp — APPROVED** [review]

File: app/module/module.go
Comment (staheri14) [inline app/module/module.go:56]: Mind explaining in the code what `fromVersion` and `toVersion` indicates for each module?

File: app/module/module.go
Comment (staheri14) [inline app/module/module.go:48]: A description for what firstVersion and lastVersion hold would be helpful.

File: app/module/module.go
Comment (staheri14) [inline app/module/module.go:60]: [question] Why the `fromVersion` is taken into account but not the `toVersion` (I guess it is related to my earlier question about the usage of fromVersion and toVersion).

File: app/module/module.go
Comment (staheri14) [inline app/module/module.go:242]: [Question] In the `EndBlock` , under a similar circumstances, where there are multiple validator updates, the method panics, but here it returns an error. Just wondering whether the two should behave similarly or not.

**Review by staheri14 — COMMENTED** [review]
LGTM, added some suggestions and questions, will defer to other reviewers for the final approval.

File: app/app.go
Comment (cmwaters) [reply app/app.go:555]: I could perhaps move this comment to the function signature. I think it's important since normally there are multiple ways of setting the consensus params and validator set that we clarify which one we use

File: app/module/module.go
Comment (cmwaters) [reply app/module/module.go:242]: This section was taken from the original module manager in the cosmos sdk

File: app/module/module.go
Comment (cmwaters) [reply app/module/module.go:60]: I don't quite understand this question. All version `fromVersion` to `toVersion` are added to the `moduleMap`

**Review by rootulp — DISMISSED** [review]

**Review by evan-forbes — DISMISSED** [review]

File: app/module/module.go
Comment (staheri14) [reply app/module/module.go:67]: What I meant to say is that the `version` variable contains the value obtained from `module.fromModule`. In the map, however, we only store the module corresponding to `fromVersion` and not `toVersion`. This leads to my confusion: shouldn't the same process apply to all versions ranging from `fromVersion` to `toVersion`? This mostly relates to my earlier question regarding the purpose of `toVersion`.

**Review by evan-forbes — APPROVED** [review]

**Review by rootulp — APPROVED** [review]

**Pull request merged**
